servers/src/main/java/tachyon/worker/block/ClientRWLock.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the University of California, Berkeley under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.worker.block;
+
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import tachyon.Constants;
+
+/**
+ * Read/write lock associated with clients rather than threads. Either its read lock or write lock
+ * can be released by a thread different from the one acquiring them (but supposed to be requested
+ * by the same client).
+ */
+public class ClientRWLock implements ReadWriteLock {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  /** Total number of permits. This value is the max number of concurrent readers */
+  private static final int MAX_AVAILABLE = 100;
+  /** Underlying Semaphore */
+  private final Semaphore mAvailable = new Semaphore(MAX_AVAILABLE, true);
+
+  @Override
+  public Lock readLock() {
+    return new UserLock(mAvailable, 1);
+  }
+
+  @Override
+  public Lock writeLock() {
+    return new UserLock(mAvailable, MAX_AVAILABLE);
+  }
+
+  private class UserLock implements Lock {
+    private final int mPermits;
+    private Semaphore mSemaphore;
```

Comment:
Can this inner class already access the outer class semaphore? If so, then would this be redundant?